pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -260,26 +266,42 @@ func (og *operationGenerator) GenerateBulkVolumeVerifyFunc(
 			}
 		}
 
-		return nil
+		return nil, nil
+	}
+
+	return volumetypes.GeneratedOperations{
+		OperationFunc: bulkVolumeVerifyFunc,
+		CompleteFunc:  util.OperationCompleteHook(pluginName, "verify_volumes_are_attached"),
 	}, nil
+
 }
 
 func (og *operationGenerator) GenerateAttachVolumeFunc(
 	volumeToAttach VolumeToAttach,
-	actualStateOfWorld ActualStateOfWorldAttacherUpdater) (func() error, string, error) {
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) (volumetypes.GeneratedOperations, error) {
 	// Get attacher plugin
+	eventRecorderFunc := func(err *error) {
+		if *err != nil {
+			for _, pod := range volumeToAttach.ScheduledPods {
+				og.recorder.Eventf(pod, v1.EventTypeWarning, kevents.FailedAttachVolume, (*err).Error())
```

Comment:
Did you verify if the value of `volumeToAttach.ScheduledPods` is populated correctly when the operation is actually executed? Want to make sure there is no funkiness in referencing a variable that is no longer in scope.

Reply:
Yep, the closure will keep the variable around. There is one nuance that we need to work through though and that is the fact that each of the "pods" in Scheduled Pods is a reference, so if the pod itself changes between the eventRecorderFunc being generated and called, it will use the new version of the pod. I think that is actually intended behavior if it is the same pod and has just been mutated slightly.

It will not, however, use any other version of "ScheduledPods" as it is passed by value so we will not have pods being added or removed in between generation and execution of this function.

I believe both of these behaviors are working as intended. lmk if you think otherwise.

Reply:
Seems like reasonable assumptions.

Reply:
I am not seeing benefit of these eventrecorder functions... The way they are being used is arbitrary too and there is a reason it is arbitrary. Depending on error each operation may want to emit one more more events (or more than one type of event) - this generalization that each operation will return a `GeneratedOperations` and that in turn will have one `EventRecorder` function simply breaks down in multiple places.

Reply:
The EventRecorder function can have arbitrary functionality to emit 0,1, or more events depending on necessity. This refactor just automates all the "plumbing" required to emit these events in one codepath. Therefore anyone wanting to emit new events or change existing events just has to add/modify the "eventRecorderFunc" of the GeneratedOperation they wish to modify.

The EventRecorder function to actual event emitted does not have to be 1 to 1.

Reply:
Discussed further with @gnufied offline. Seems like the issue is that this abstraction only has the ability to capture "Error" type events that are returned from the generated function. The abstraction "breaks down" when we try to use it for other types of events such as a Warning or Success.

The intended design of the abstraction was only for Error events, while warnings and success events would continue to use the direct og.recorder.Eventf method. This can be documented by changing the EventRecorderFunc to ErrorEventRecorderFunc so that it is no longer a misnomer.

@gnufied please confirm or correct these statements per your understanding.

Reply:
hmm - I am not sure about net benefit of having this recorder in the first place. I have a use case - where I want to reduce severity of an event depending on error raised. I still want the whole operation to fail in general (because a non-failing operation will not respect exponential backoff) but in certain cases i want the error to be a warning (because operation will be retried anyways).